keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,316 @@
+
+### Installed Operators
+
+Before an upgrade can be performed, the CLI must validate that all installed operators are compatible with the new KUDO version.
+To do this, each operator has certain attributes:
+- `kudoVersion` This determines the minimum version of KUDO that is required for this operator
+- `apiVersion` This defines the package format of the operator
+  - Currently both `operator.yaml` and `params.yaml` have an `apiVersion`, and they should be kept in sync, even if they could be changed separately
+  - The `apiVersion` is unrelated to the CRD version - the CLI converts from the package format to the CR, but the versions can change independently
+
+As the `apiVersion` (and the package format) is only used to install new operatorVersions, it does not play into the compatibility concerns when upgrading
+KUDO. 
+
+TODO: Add more details
+
+### KUDO Prerequisites
+Expected update frequency: Medium  
+Versioned: No, but closely tied to KUDO manager
+
+The KUDO manager has a set of prerequisites that need to be present for the manager to run successfully. They are
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+Usually the prerequisites are API resources, but may be more complex things.
+
+- Prerequisites may be feature-gated
+- Prerequisites may be k8s-version-dependent
+- Prerequisites may have other prerequisites as dependencies (i.e., service account needs the namespace to exist)
+- Prerequisites may have parameters and behave differently based on them (i.e. default namespace may be created, but a provided namespace needs to exist)
+
+- Possible Prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secrets
+  - Webhooks
+  - Other software in the cluster (i.e. cert-manager)
+
+- For each prerequisite, there are a finite set of possible options in an update case:
+  - The Prereq exists and has the same version/content as the new one - no action required
+  - The Prereq does not exist - needs to be installed/created
+  - The Prereq exists and has different version/content - needs to be updated/replaced
+  - (The Prereq exists but should not exist anymore - needs to be removed/deleted) 
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+At the current time, it seems enough to use a simple install/update process.
+  - The setup/update contains a list of all prerequisites in correct order
+  - Each prereq validates the current installed state, and verifies that it can install/update the current state to the expected state
+  - Prereqs that are deleted in newer versions need to stay in the list of prerequisites
+
+We may arrive at a point where we need to implement custom migration logic between KUDO versions, but so far this would mean a lot of
+implementation effort without a defined use case.    
+
+### KUDO Manager
```

Comment:
Is there any expectation on the manager to re-process plans that have been run on previous versions?  I'm thinking of scenarios where there might be a bug in a previous version that would have caused the plan to not be executed correctly, and want to re-process the plans to correct the mistake.

Reply:
I don't think that would be a good idea to do automatically. For some plans that may be ok, but others not as much - restore plans, node replaces, etc. 

There will be some kind of framework to write migrations, I guess, to fix major issues that need to be fixed, but nothing like just re-processing plans.